### ISSUE TITLE
Return instructions from the resolver

### ISSUE THREAD

**Issue by kettanaito** [kettanaito/msw#1207]
### Scope

Improves an existing behavior

### Compatibility

- [ ] This is a breaking change

### Feature description

Instead of a response resolver returning a `MockedResponse` object, I propose it returns an abstract instruction object. Such an instruction object tells the library what to do next with the intercepted request.

- Originates from #1204 

## Instruction types

- Fallthrough. Ignore this handler, skip to the next matching handler, if any.
- Passthrough (see #1204). Perform the intercepted request as-is.
- Use mocked response (currently `res()`).
- Return a network error (currently `res.networkError()`).
- Any arbitrary custom instruction in the future to be handled by `handleRequest`.

## Motivation

There are multiple sources that may affect the resolution of an intercepted request. You may compose a mocked response via `res()`, return a network error, or instruct the library to explicitly perform this request as-is. Pushing all of those different intentions into a single `MockedResponse` object is not maintainable. 

Even the behaviors like `once` do not belong on the mocked response object. Instead, things like `once` must be _modifiers_ to instructions, not even instructions themselves. This will allow us to return any instructions with a modifier, such as "return a network error once" (see #413).

**Comment by Joshuaweiss** [comment]
I'm using a wildcard resolver to check all requests authorization tokens against a fixed list of mock session tokens and want to fall through if the token is valid. Returning undefined from the function seems to accomplish this although it produces a warning without explicit req.fallthrough support.

**Comment by kettanaito** [comment]
@Joshuaweiss, we've introduced `req.passthrough()` as an explicit way to achieve just that and don't get the warning.

```js
rest.get('/resource', (req) => {
  return req.passthrough()
})
```

I believe it simply returns a `101 Continue` response, which MSW treats as "okay, let's keep searching for the relevant handler".

**Comment by Joshuaweiss** [comment]
It seems like `req.passthrough()` bypasses mocking entirely for that request. I want the request to fallthrough to a later handler which seems to be what happens when you return `undefined` but that also produces a warning.

I'm trying to use this to imitate an auth middleware which check auth header before any specific mocked endpoint handler.

**Comment by kettanaito** [comment]
Yes, you are right, @Joshuaweiss. I think that the fallthrough functionality is still extremely confusing that I myself have mixed up what does what. 

The expectation behind `req.passthrough()` to stop the handler lookup and perform the request as-is _is correct_.

**Comment by kettanaito** [comment]
I've refactored the request handler lookup logic in #1608 and will consider this issue solved once it gets merged. **The changes will be available in the next major release**. 

Here's what you will be able to do with a response resolver then:

## Option 1: Return a mocked response

```js
rest.get('/user', () => {
 return new HttpResponse('Hello world!')
})
```

## Option 2: Bypass a request

```js
import { passthrough } from 'msw'

rest.get('/user', () => {
 return passthrough()
})
```

> Under the hood, the `passthrough()` function still complies with Option 1 and returns a `302` response redirect to produce a semantic response that's handled by MSW internally. 

This _stops the request lookup_ and performs the request as-is. This function is useful to capture requests you know you want to explicitly perform as-is. 

## Option 3: Fallthrough to other handlers

```js
rest.get('*', () => {
  return
}),
rest.get('/user', () => new HttpResponse('Hello world!'))
```

Dispatching a `GET /user` request will:

1. Go to the first `rest.get('*')` handler because it matches all `GET` requests.
2. See that handler returned nothing, so it will continue looking for a handler that returns something.
3. It will find the next `rest.get('/user')` handler and since it returns a mocked response, it will use that mocked response. 

The issue will continue until either of the following criteria are met:

1. There are no more request handlers.
2. There's a matching request handler that produced a response.

**Comment by Joshuaweiss** [comment]
This looks great, Thanks!

**Comment by kettanaito** [comment]
Based on the discoveries from #1671, you cannot differentiate in TypeScript between these two functions:

```ts
function a() {}

function b() {
 return 
}
```

They both are described as `() => void` by TypeScript. This makes it impossible for us to establish the return type contract I described earlier because we can't guard against "missing return" scenarios. 

To address this, I think we should introduce a new function that means "fallthrough()", and instructs MSW to continue the request lookup until the next matching handler is found. The `fallthrough()` function will be a substitute for "return", and, internally, will also return a `Response` instance describing the fallthrough intention.

**Comment by Joshuaweiss** [comment]
@kettanaito Personally I think the explicit `fallthrough` function is the better interface and the one I would expect to see. I went to look for a function like `passthrough` and then looked at the return type for the resolver function first. I only saw the `undefined` would work as a return type by looking at the implementation and was pleasantly surprised to find I could use `undefined` to sides step my issue. Returning undefined can be ambiguous and is no different at runtime from not returning anything at all. I believe ultimately TypeScript tries to as close as possible reflect runtime behavior so it makes sense they don't differentiate.

**Comment by kettanaito** [comment]
@Joshuaweiss, thanks for your input! I thought about this and came to the conclusion that default implicit behaviors are fine. I'm not a fan of implicit at all but certain things work well when unspecified. The fallthrough scenario is one of those things. 

The problem here is that introducing mandatory explicit instruction from the resolver would inevitably feel like a chore to the user. Right now, MSW works on the premise "we intercept requests but will do nothing until you tell us to". With the explicit `fallthrough()`, that premise will change to "we intercept requests but will error until you tell us what to do." Those are quite different ideological approaches to mocking, and I dislike the intrusive nature of the second one. 

The only practical drawback to that implicit fallthrough is that it's hard to guard against false positive intentions on build- on runtime. As described in detail #1671, TypeScript doesn't have the means to enforce _any_ return on a function. However, in the same issue, I explain why this isn't a huge concern the library should have. 

Apart from that, implicit fallthrough gives a great advantage to every user, also making it easier to learn how to work with the library.

**Comment by kettanaito** [comment]
## Released: v2.0.0 🎉

This has been released in v2.0.0!

- 📄 [**Release notes**](https://github.com/mswjs/msw/releases/tag/v2.0.0)
- 📦 [npm package](https://www.npmjs.com/package/msw/v/2.0.0)

Make sure to always update to the latest version (`npm i msw@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.
